components/resource_control/src/worker.rs

```diff
@@ -0,0 +1,430 @@
+            .resource_ctl
+            .resource_groups
+            .iter()
+            .filter_map(|kv| {
+                let g = kv.value();
+                g.limiter.as_ref().map(|limiter| GroupStats {
+                    name: g.group.name.clone(),
+                    ru_quota: g.get_ru_quota() as f64,
+                    limiter: limiter.clone(),
+                    stats_per_sec: GroupStatistics::default(),
+                    expect_cost_rate: 0.0,
+                })
+            })
+            .collect();
+        if background_groups.is_empty() {
+            return;
+        }
+
+        self.do_adjust(ResourceType::Cpu, dur_secs, &mut background_groups);
+        self.do_adjust(ResourceType::Io, dur_secs, &mut background_groups);
+    }
+
+    fn do_adjust(
+        &mut self,
+        resource_type: ResourceType,
+        dur_secs: f64,
+        bg_group_stats: &mut [GroupStats],
+    ) {
+        let resource_stats = match self.resource_quota_getter.get_current_stats(resource_type) {
+            Ok(r) => r,
+            Err(e) => {
+                warn!("get resource statistics info failed, skip adjust"; "type" => ?resource_type, "err" => ?e);
+                return;
+            }
+        };
+        // if total resource quota is unlimited, set all groups' limit to unlimited.
+        if resource_stats.total_quota <= f64::EPSILON {
+            for g in bg_group_stats {
+                g.limiter
+                    .get_limiter(resource_type)
+                    .set_rate_limit(f64::INFINITY);
+            }
+            return;
+        }
+
+        let mut total_ru_quota = 0.0;
+        let mut background_consumed_total = 0.0;
+        let mut has_wait = false;
+        for g in bg_group_stats.iter_mut() {
+            total_ru_quota += g.ru_quota;
+            let total_stats = g.limiter.get_limiter(resource_type).get_statistics();
+            let mut stats_delta =
+                match self.prev_stats_by_group[resource_type as usize].entry(g.name.clone()) {
+                    Entry::Occupied(mut s) => total_stats - s.insert(total_stats),
+                    Entry::Vacant(v) => {
+                        v.insert(total_stats);
+                        total_stats
+                    }
+                };
+            stats_delta = stats_delta / dur_secs;
+            background_consumed_total += stats_delta.total_consumed as f64;
+            g.stats_per_sec = stats_delta;
+            if stats_delta.total_wait_dur_us > 0 {
+                has_wait = true;
+            }
+        }
+
+        // fast path if process cpu is low
+        let is_low_load = resource_stats.current_used <= (resource_stats.total_quota * 0.1);
+        if is_low_load && !has_wait && self.is_last_time_low_load[resource_type as usize] {
+            return;
+        }
+        self.is_last_time_low_load[resource_type as usize] = is_low_load;
+
+        let mut available_resource_rate = ((resource_stats.total_quota
```

Comment:
please add some comments here about how we decide the available resource for background

Reply:
done.